File: leetcode/0337-house-robber-iii/solution.cpp
#include <algorithm>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int rob(TreeNode* root) {
        auto [robbed, notRobbed] = helper(root);
        return max(robbed, notRobbed);
    }

private:
    pair<int, int> helper(TreeNode* node) {
        if (node == nullptr) {
            return {0, 0};
        }

        auto [leftRobbed, leftNot] = helper(node->left);
        auto [rightRobbed, rightNot] = helper(node->right);

        int robbed = node->val + leftNot + rightNot;
        int notRobbed = max(leftRobbed, leftNot) + max(rightRobbed, rightNot);

        return {robbed, notRobbed};
    }
};

File: leetcode/0337-house-robber-iii/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def helper(node: Optional[TreeNode]):
            if node is None:
                return (0, 0)  # (robbed, not_robbed)

            left_robbed, left_not = helper(node.left)
            right_robbed, right_not = helper(node.right)

            robbed = node.val + left_not + right_not
            not_robbed = max(left_robbed, left_not) + max(right_robbed, right_not)

            return (robbed, not_robbed)

        return max(helper(root))
